draft-reddy-ipsecme-ikev2-pqc-auth.md

```diff
@@ -131,6 +132,53 @@ SLH-DSA does not introduce a new hardness assumption beyond those inherent to th
+The resulting signature is placed into the The resulting signature is the Signature Value.
+
+When verifying a signature with ML-DSA or SLH-DSA, the IKEv2 implementation would take the InitiatorSignedOctets string or the ResponderSignedOctets string (as appropriate), logically send it to the identity hash (which leaves it unchanged), and then pass it into the ML-DSA or SLH-DSA signer as the message to be verified (along with the fixed context string "IKEv2 Auth".
+
+## Discuession of ML-DSA and SLH-DSA and Prehashing
+
+The signature architecture within IKE was designed around RSA (and later extended to ECDSA).
+In this architecture, the actual message (the SignedOctets) are first hashed (using a hash that the verifier has indicated support for), and then passed for the remaining part of the signature generation processing.
+That is, it is designed for signature algorithms that first apply one of a number of hash functions to the message and then perform processing on that hash.
+Neither ML-DSA nor SLH-DSA fits cleanly into this architecture.
+
+We see three ways to address this mismatch.
+
+The first is to note that both ML-DSA and SLH-DSA have prehashed parameter sets; that is, ones designed to sign a message that has been hashed by an external source.
+At first place, this would appear to be an ideal solution, however it turns out that there are a number of practical issues.
+The first is that the prehashed version of ML-DSA and SLH-DSA would appear to be rarely used, and so it is not unlikely that support for it within crypto libraries may be lacking.
+The second is that the public keys for the prehashed versions use different OIDs; this means that the certificates for IKEv2 would necessarily be different than certificates for other protocols (and some CAs might not support issuing certificates for prehashed ML-DSA or prehashed SLH-DSA, again because of the lack of use).
+The third is that some users have expressed a desire not to use the prehashed parameter sets.
+
+The second is to note that, while IKEv2 normally acts this way, it doesn't always.
+EdDSA has a similar constraint on not working cleanly with the standard 'hash and then sign' paradigm, and so the existing [RFC8420] provides an alternative method, which ML-DSA would cleanly fit into.
+We could certainly adopt this same strategy; our concern would be that it might be more difficult for IKEv2 implementors which do not already have support for EdDSA.
+
+The third way is what we can refer to as 'fake prehashing'; IKEv2 would generate the hash as current, but instead of running ML-DSA or SLH-DSA in prehash mode, we have itsign it in pure mode as if it was the message.
+This is a violation of the spirit, if not the letter of FIPS 204, 205
```

Comment:
we should considering discussing the use of ExternalMu-ML-DSA, see https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/. It has an advantage that IKEv2 would not have to generate the hash and ExternalMu-ML-DSA will first compute the hash for the message and then sign the hash. It can also be used with HSMs (where ExternalMu-ML-DSA.Prehash is computed in the software cryptographic module and ExternalMu-ML-DSA.Sign in the HSM).

Reply:
Actually, ExternalMu-ML-DSA would be equivalent to the RFC 8420 approach (and in essence, another way of implementing it).  However.
- Not all cryptolibraries (HSMs) can implement it (sorry, but some HSMs follow the APIs that NIST defines and nothing else).
- It doesn't work for SLH-DSA (that is, malicious software could trick the HSM into signing messages that would make forgery of arbitrary messages feasible). 

With those caveats, I suppose we could mention ExternalMU-ML-DSA as a possible implementation option...

Reply:
The third method has another drawback: a hash algorithm used at the IKEv2 level to hash the message needs to be specified for interoperability, which contradicts the 'Identity' hash in the prior section.

It seems that the second approach faces the least resistance, and an implementation can use ExternalMU-ML-DSA or the pure mode without causing any interoperability issues.

We will have to mandate the use of the second method.

Reply:
As for the third method contracting the 'identity hash', that third method is an alternative way of integrating ML-DSA into IKEv2, not a different implementation technique.  We wouldn't use an identity hash - instead, we would use a negotiated hash, and then sign that hash.  So, the  change from what IKEv2 does now is replacing 'most of RSA, that is, everything except for the hash' with 'all of ML-DSA'.

Reply:
Added ExternalMU-ML-DSA text.  I couldn't add a reference to where ExternalMU-ML-DSA was defined (I believe it's in https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/07/; it's not in FIPS 204)